Fixes #36711 - Fix searching in settings

girijaasoni opened the pull request:
This PR addresses: 

- Allowing description search with fuzzy operator
-  Displaying all the relevant suggestions for autocomplete while searching with name including the suggestions for settings with Non Nil values.

stejskalleos commented:
ok to test

nofaralfasi reviewed (changes requested):
I have a few observations to share regarding this PR:
As I discussed previously with @girijaasoni in a private conversation, there are a couple of issues that require attention:
- The 'and'/'or' operations don't seem to be functioning correctly.
- When searching for descriptions containing special characters, it triggers a 500 error code or generates an error message.

One more note, I'm not sure why we didn't opt for the ext_method option, which appears to address our concerns, as detailed in this link: [GitHub PR #178](https://github.com/wvanbergen/scoped_search/pull/178). Given that the description search is already operational through the code found here: [Setting Presenter](https://github.com/theforeman/foreman/blob/b9986c9648b55c0d8fae4ab5ff5845763ddbea1f/app/presenters/setting_presenter.rb#L81), I'm uncertain about the need for additional code. Wouldn't it be more efficient to resolve the error by incorporating the ext_method into the settings model?

girijaasoni commented:
> I have a few observations to share regarding this PR: As I discussed previously with @girijaasoni in a private conversation, there are a couple of issues that require attention:
> 
> * The 'and'/'or' operations don't seem to be functioning correctly.
> * When searching for descriptions containing special characters, it triggers a 500 error code or generates an error message.
> 
> One more note, I'm not sure why we didn't opt for the ext_method option, which appears to address our concerns, as detailed in this link: [GitHub PR #178](https://github.com/wvanbergen/scoped_search/pull/178). Given that the description search is already operational through the code found here: [Setting Presenter](https://github.com/theforeman/foreman/blob/b9986c9648b55c0d8fae4ab5ff5845763ddbea1f/app/presenters/setting_presenter.rb#L81), I'm uncertain about the need for additional code. Wouldn't it be more efficient to resolve the error by incorporating the ext_method into the settings model?

Thanks @nofaralfasi 

Searching in the search bar is implemented by the search_for function which gives us the results, it has always been operational. However, in this PR, The concern addressed is implementing the autocomplete functionality which needs some existing scoped search functions to be overridden. IMHO,  The ext method is usually used in case of searching from a different table column or searching based on creating SQL relations as per this  [Reference](https://github.com/wvanbergen/scoped_search/wiki/Search-definition#external-search-methods) and not when we are searching from a memory object.

stejskalleos reviewed (commented):
I have a different idea:

How about turning off suggestions completely, and do not show `name` and `description`  in the form at all?
The only thing users could do would be to enter value into the search form, and that's it.
The search would then be equal to the `name ~ $VALUE OR description ~ $VALUE`

It's simple and it's going to cover 99% cases IMHO. It looks to me like it's unnecessarily complex for a simple search in a very simple table.

nofaralfasi commented:
> It's simple and it's going to cover 99% cases IMHO. It looks to me like it's unnecessarily complex for a simple search in a very simple table.

I agree, and we can always add this functionality if users request it.

girijaasoni commented:
> I have a different idea:
> 
> How about turning off suggestions completely, and do not show `name` and `description` in the form at all? The only thing users could do would be to enter value into the search form, and that's it. The search would then be equal to the `name ~ $VALUE OR description ~ $VALUE`
> 
> It's simple and it's going to cover 99% cases IMHO. It looks to me like it's unnecessarily complex for a simple search in a very simple table.

Turning off auto complete would break the search pattern from the other components. In my opinion, suggestions help the user the determine what exactly they want to look for.  In this PR, autocomplete works and is only an issue when there's the use of special characters which is fixable(I will update it soon). I think  removing the name and description search would be a great hack if and when it's not implementable (which is not the case here) and might require additional suggestions and votes.

stejskalleos commented:
> Turning off auto complete would break the search pattern from the other components.

Is it an issue? This fix looks like overkill to me.

ofedoren commented on app/controllers/api/v2/settings_controller.rb:
```diff
@@ -22,7 +22,6 @@ class SettingsController < V2::BaseController
 
       api :GET, "/settings/", N_("List all settings")
       param_group :search_and_pagination, ::Api::V2::BaseController
-      add_scoped_search_description_for(Setting)
```
Please, don't remove it completely, it's for documenting what can be used for searching via API/hammer.

ares commented:
A few notes from the testing:
* this makes searching by description possible again, with the `~` operator. It gives auto completion list though, which I would not expect as it shows complete descriptions.
* for the name it does not work at all, the auto completion suggest the names that user sees in the table (good), these can have spacebars but the value is not wrapped in quotes (bad), even if I add quotes it does not work (bad), because the search actually responds on the setting identifier. So e.g. `name =  "Foreman URL"` does not work while `name = foreman_url` works fine.
* searching on name also auto suggests only `~`, while other operators like `=` works fine

ares commented:
To illustrate I recorded few videos.

[description.webm](https://github.com/theforeman/foreman/assets/109773/7819d308-153e-4ac0-8947-e51dd1ce0a9f)

description shows auto completion, that feels overwhelming and does not add quotes around the multiword term

[name_exact.webm](https://github.com/theforeman/foreman/assets/109773/d56221d2-f871-435d-87c2-b4f28f832b7f)

name with typing equal sign provides full name sugestions that don't work at all, even with manually added quotes, as it turns out, it searches by name, e.g. `Foreman URL` vs `foreman_url`

[name_fuzzy.webm](https://github.com/theforeman/foreman/assets/109773/e9b24f0e-ea67-4c1d-92a3-9d35d01ee553)

fuzzy searching on name seems to work (not sure why `=` is not autosuggested), but again, does not add quotes, so "Foreman URL" will give you all settings with either Foreman or URL in their name. With added quotes, it searches correctly.

Hope this helps. Again to be clear, searching in full name is preferred over the name. I would love to search for `Foreman URL` instead of `foreman_url`, provided it works.

girijaasoni commented:
> To illustrate I recorded few videos.
> 
>  description.webm 
> description shows auto completion, that feels overwhelming and does not add quotes around the multiword term
> 
>  name_exact.webm 
> name with typing equal sign provides full name sugestions that don't work at all, even with manually added quotes, as it turns out, it searches by name, e.g. `Foreman URL` vs `foreman_url`
> 
>  name_fuzzy.webm 
> fuzzy searching on name seems to work (not sure why `=` is not autosuggested), but again, does not add quotes, so "Foreman URL" will give you all settings with either Foreman or URL in their name. With added quotes, it searches correctly.
> 
> Hope this helps. Again to be clear, searching in full name is preferred over the name. I would love to search for `Foreman URL` instead of `foreman_url`, provided it works.

> To illustrate I recorded few videos.
> 
>  description.webm 
> description shows auto completion, that feels overwhelming and does not add quotes around the multiword term
> 
>  name_exact.webm 
> name with typing equal sign provides full name sugestions that don't work at all, even with manually added quotes, as it turns out, it searches by name, e.g. `Foreman URL` vs `foreman_url`
> 
>  name_fuzzy.webm 
> fuzzy searching on name seems to work (not sure why `=` is not autosuggested), but again, does not add quotes, so "Foreman URL" will give you all settings with either Foreman or URL in their name. With added quotes, it searches correctly.
> 
> Hope this helps. Again to be clear, searching in full name is preferred over the name. I would love to search for `Foreman URL` instead of `foreman_url`, provided it works.

Thanks a lot @ares 
These videos were really helpful

In this push I have addressed these few things:

-  Ditched autocomplete  for description fuzzy operator as they were very long
-  name = "Foreman URL" or name = Foreman URL wasn't working as we didn't define searching with full name in our matches_search_query method which was unrelated to scoped search but since it was a minor fix so I added it in this PR, Now we can search for name = "Foreman URL"
- Wrapped the name suggestions with quotes for multiple words

ares commented:
From user perspective, this works now as expected. There's one issue, if user enters something that's not supported, e.g. ` `id =`, then they see Error: Request failed with status code 500.

![settings](https://github.com/theforeman/foreman/assets/109773/5b6c8420-debc-4071-afc1-3579cb2aec79)

Previously it was showing better error message.

![set2](https://github.com/theforeman/foreman/assets/109773/a80482e5-4095-4820-a78d-6fddc7b343e9)

This should be fixed I guess

ShimShtein commented on app/services/setting_registry.rb:
```diff
@@ -1,6 +1,62 @@
+    end
+  end
+
+  scoped_search :on => :id, :complete_enabled => false, :only_explicit => true, :validator => ScopedSearch::Validators::INTEGER
+  scoped_search on: :name, complete_value: :true, operators: ['~', "="]
+  scoped_search on: :description, complete_value: :true, operators: ['~']
```
Those definitions should be moved to [setting model](https://github.com/theforeman/foreman/blob/develop/app/models/setting.rb#L57), since the scoped search does not use the `SettingRegistry` class for search definition.

girijaasoni replied:
we have the definitions in the setting model as well for hammer purpose, but we need it in the registry as well because we need to pass the scoped search definition in the auto_complete method (which we are overriding)

ShimShtein replied:
That's not a problem, just pass the definition from the `Setting` model:
```
    SettingCompleter.auto_complete(self, Setting.scoped_search_definition, query, opts)
```
Otherwise we are creating duplication of search definitions, and it will be hard to track down which definition is used.

girijaasoni replied:
That was a great catch @ShimShtein 
updated it :)

ShimShtein commented on app/services/setting_registry.rb:
```diff
@@ -1,6 +1,62 @@
+      elsif field.field == :description
+        results = []
+      else
+        raise ::Foreman::Exception.new N_('Unsupported completion, only name and description are supported')
```
We have to send here a `ScopedSearch::QueryNotSupported` error instead of a general `Exception`, so the error would be handled correctly by the controller.

girijaasoni replied:
THanks @ShimShtein 
that's doable

girijaasoni commented:
> From user perspective, this works now as expected. There's one issue, if user enters something that's not supported, e.g. ` `id =`, then they see Error: Request failed with status code 500.
> 
> ![settings](https://user-images.githubusercontent.com/109773/281433698-5b6c8420-debc-4071-afc1-3579cb2aec79.png)
> 
> Previously it was showing better error message.
> 
> ![set2](https://user-images.githubusercontent.com/109773/281434542-a80482e5-4095-4820-a78d-6fddc7b343e9.png)
> 
> This should be fixed I guess



> From user perspective, this works now as expected. There's one issue, if user enters something that's not supported, e.g. ` `id =`, then they see Error: Request failed with status code 500.
> 
> ![settings](https://user-images.githubusercontent.com/109773/281433698-5b6c8420-debc-4071-afc1-3579cb2aec79.png)
> 
> Previously it was showing better error message.
> 
> ![set2](https://user-images.githubusercontent.com/109773/281434542-a80482e5-4095-4820-a78d-6fddc7b343e9.png)
> 
> This should be fixed I guess

test

girijaasoni commented:
> From user perspective, this works now as expected. There's one issue, if user enters something that's not supported, e.g. ` `id =`, then they see Error: Request failed with status code 500.
> 
> ![settings](https://user-images.githubusercontent.com/109773/281433698-5b6c8420-debc-4071-afc1-3579cb2aec79.png)
> 
> Previously it was showing better error message.
> 
> ![set2](https://user-images.githubusercontent.com/109773/281434542-a80482e5-4095-4820-a78d-6fddc7b343e9.png)
> 
> This should be fixed I guess

Thanks @ares 
Fixed it in this commit
@stejskalleos , I also cutshort the complete_value method (the method which was causing the overkill) as much as possible

ShimShtein commented on app/services/setting_registry.rb:
```diff
@@ -1,6 +1,49 @@
+    def complete_value_from_db(field, val)
+      count = 20
+      if field.field == :name
+        results = @registry.map { |set| ((set.full_name =~ /\s/) ? "\"#{set.full_name.gsub('"', '\"')}\"" : set.full_name) if set.name.include?(val) || set.full_name&.include?(val) }
```
This will return empty values when the condition is not met. e.g:
For `settings => ['aaa', 'abc', 'bbb']` and `val => 'b'` you will get:
`[nil, 'abc', 'bbb']`.
We should use [`filter_map`](https://www.rubydoc.info/stdlib/core/Enumerable#filter_map-instance_method) instead:
```suggestion
        results = @registry.filter_map { |set| ((set.full_name =~ /\s/) ? "\"#{set.full_name.gsub('"', '\"')}\"" : set.full_name) if set.name.include?(val) || set.full_name&.include?(val) }
```

girijaasoni replied:
That makes much more sense, thanks!

girijaasoni committed:
Fixes #36711 - Fix searching in settings

ShimShtein reviewed (approved):
LGTM, but I don't have the latest Foreman to test. Can someone take this one to a quick run?

stejskalleos reviewed (approved):
Tested:
* Default search
* `name ~` & `name =`
* `description`

Looks like it's working, let's get this in!

stejskalleos commented:
Thanks @girijaasoni, @ezr-ondrej, @ShimShtein
